src/test/java/com/flightstats/hub/util/IntegrationUdpServer.java

```diff
@@ -1,74 +1,71 @@
 package com.flightstats.hub.util;
 
 import lombok.Builder;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import lombok.extern.slf4j.Slf4j;
 
 import java.io.BufferedReader;
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.net.DatagramPacket;
 import java.net.DatagramSocket;
 import java.io.IOException;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
 import java.util.stream.Collectors;
 
+@Slf4j
 @Builder
 public class IntegrationUdpServer {
-    private final static Logger logger = LoggerFactory.getLogger(IntegrationUdpServer.class);
-    private long timeoutMillis;
-    private final List<String> queue = new ArrayList<>();
     private int port;
     private boolean listening;
+    private CountDownLatch startupCountDownLatch;
```

Comment:
All of these instance variables, are they used across threads? None are volatile or wrapped in AtomicReference so there'd be potential race conditions or lack of cross-thread visibility guarantees.

Reply:
Would it be acceptable to turn these into parameters on the method that uses them and make them final there instead? The test that uses this creates them as final members... I don't have a great grasp of final vs volatile when it comes to visibility and mutations.

Reply:
Either final or volatile will work. Final variables are guaranteed to be visible once construction is finished, since they're write once the JVM notifies all other threads of the final value. Volatile would also work for visibility because that tells the JVM to always check the value of the variable and never rely on what the thread has cached.

Final is going to be more performant if that matters here. It's also simpler to reason about since something that never changes can't have race conditions. I'd suggest making "listening" ```final AtomicBoolean```.

If you do any multi-thread work, it's definitely worth reading the chapter about visibility in Java Concurrency in Practice.

Reply:
Edit: Oh, you asked about final method parameters? Method parameters don't have concurrency concerns, other than the things you pass could themselves be used in other threads.

```final``` is pretty pointless for method parameters or variables in a method. It just prevents you from assigning a new value to that parameter...which is generally bad form anyway (not always, but usually). ```final``` for method parameters doesn't have any impact for concurrency. I don't think ```volatile``` even applies for method parameters, I believe that's only for instance variables.

Reply:
Right. Java uses "Pass by value" for method parameters.

Reply:
I wound up moving all those variables out of the class, so I guess my question is moot. I'll take a look at the concurrency book, though. Thanks!